### DISCUSSION TITLE
Is there an idiomatic way to attach data to a number of different assertions?

### DISCUSSION THREAD

**Discussion by anthony-c-martin** [fluentassertions/fluentassertions#1509] [Q&A]
## Scenario
Imagine something like the below scenario, where I'd like the additionalInfo to be logged along with any of the assertions if they fail.

```csharp
var additionalInfo = ExpensiveMethodToPrintDebugInfo());

a.Should().Be(b); // on assertion failure, the message should include additionalnfo
b.Should().NotBeNull(); // on assertion failure, the message should include additionalnfo
...many more assertions...
```

Some approaches I already looked at below. I couldn't quite figure out how to make them fit. Sorry if I'm missing something obvious!

### Using the AssertionScope
* Almost does what I need, but also affects the call flow - I want to retain the behavior of failing the test on the first assertion.
* AddReportable forces me to calculate the value up front instead of deferring it to only be calculated on failure. This is expensive, as I have tests which loop frequently, and want to attach detailed (and expensive to calculate) debugging info to the assertion message.
```csharp
using (var assertionScope = new AssertionScope())
{
    assertionScope.AddReportable("additionalInfo", ExpensiveMethodToPrintDebugInfo());
    
    a.Should().Be(b);
    b.Should().NotBeNull();
    // I'd like to be able to use 'b' as if it were non-null here onwards, but the modified call flow means 'b' could still be null.
    ...many more assertions...
}
```

### Using a ValueFormatter
* As far as I could tell, this only gives me the ability to format the message based on properties of the object I'm asserting on. In my scenario, I need to use a bunch of other contextual properties to format the message, so this feels non-viable.

### Creating my own assertions
I considered whether I could just follow https://fluentassertions.com/extensibility/#building-your-own-extensions and add a bunch of assertions which took in the additional context - e.g.:
```csharp
a.Should().Be(b, ExpensiveMethodToPrintDebugInfo());
b.Should().NotBeNull(ExpensiveMethodToPrintDebugInfo());
```
* This feels hacky, plus I'd have to reimplement every single one of the existing assertions that I want to use to take my debugging context.

**Comment by dennisdoomen** [comment]
Other than using the `because` argument all assertions have or the `AssertionScope` you suggested, I can't think of anything to make that possible.

**Reply by lg2de** [reply]
I see the following options:
1. add overload taking `Func<string>` instead of `string because`.
2. add handling the type `Func<string>` in the list of `becauseArgs`
3. add overload taking `Func<string>` to `AssertionScope.AddReportable`
